### PR TITLE
add basic system tests for OpenLineage

### PR DESCRIPTION
**Pull request by mobuchowski**
Added system tests to make sure our OpenLineage integration actually works as expected. Here's what's new:

- Built a VariableTransport that dumps OpenLineage events into Airflow Variables so we can check them later
- Created a test operator (OpenLineageTestOperator) that compares the events we got against what we expected
- Added a simple test DAG to show how it all works together
- Threw in some helper functions to make event validation less painful (checking UUIDs, dates, etc.)

The main idea is to have a reliable way to verify that we're sending the right OpenLineage events with the right format.
The test setup is pretty flexible - you can either define your expected events inline or keep them in JSON files, whatever works better for your use case.

### CODE REVIEW

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    :raises: ValueError if the received events do not match with expected ones.
+    """
+
+    def __init__(
+        self,
+        event_templates: dict[str, dict] | None = None,
+        file_path: str | None = None,
+        env: Environment = setup_jinja(),
+        multiple_events: bool = False,
+        delete: bool = True,
```
Comment (kacpermuda) [inline providers/src/airflow/providers/openlineage/utils/operator.py]: These two arguments are missing in the docstring

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        self.multiple_events = multiple_events
+        self.delete = delete
+        if self.event_templates and self.file_path:
+            raise ValueError("Can't pass both event_templates and file_path")
+        if self.file_path is not None:
+            self.event_templates = {}
+            with open(file_path) as f:  # type: ignore[arg-type]
+                events = json.load(f)
+            for event in events:
+                key = event["job"]["name"] + ".event." + event["eventType"].lower()
```
Comment (kacpermuda) [inline providers/src/airflow/providers/openlineage/utils/operator.py]: Shouldn't we extract this key creation logic somewhere since it's used in two places to avoid potential discrepancies in the future?

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 1 above)
Comment (kacpermuda) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
        clear_variables: bool = True,
```
Maybe some more descriptive name can be used here?

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                self.event_templates[key] = event
+
+    def execute(self, context: Context):
+        for key, template in self.event_templates.items():  # type: ignore[union-attr]
+            send_event = Variable.get(key=key, deserialize_json=True)
+            if len(send_event) != 1 and not self.multiple_events:
+                raise ValueError(f"Expected one event for key {key}, got {len(send_event)}")
+            if len(send_event) == 0:
+                raise ValueError(f"No event for key {key}")
```
Comment (kacpermuda) [inline providers/src/airflow/providers/openlineage/utils/operator.py]: If there are no events the first condition will catch that when `multiple_events` is `False` (default), maybe we should reorder those ifs?

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    :param env: jinja environment used to render event templates
+    :raises: ValueError if the received events do not match with expected ones.
+    """
+
+    def __init__(
+        self,
+        event_templates: dict[str, dict] | None = None,
+        file_path: str | None = None,
+        env: Environment = setup_jinja(),
+        multiple_events: bool = False,
```
Comment (kacpermuda) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
        allow_duplicate_events: bool = False,
```
Maybe we can come up with something more descriptive?

File: providers/tests/system/openlineage/conftest.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@pytest.fixture(autouse=True)
+def set_transport_variable():
+    lm = get_listener_manager()
+    lm.clear()
+    listener = OpenLineageListener()
+    listener.adapter._client = listener.adapter.get_or_create_openlineage_client()
+    listener.adapter._client.transport = VariableTransport()
+    lm.add_listener(listener)
+    yield
+    lm.clear()
```
Comment (kacpermuda) [inline providers/tests/system/openlineage/conftest.py:36]: We are removing all listeners and not re-setting them after yield. There are probably no listeners anyway, but just wanted to confirm it's intended here.

File: providers/src/airflow/providers/openlineage/transport/variable.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    It's made to be used in system tests, stored data read by OpenLineageTestOperator.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        ...
+
+    def emit(self, event: Event):
+        key = f"{event.job.name}.event.{event.eventType.value.lower()}"  # type: ignore[union-attr]
+        self.log.error("SETTING KEY %s", key)
```
Comment (kacpermuda) [inline providers/src/airflow/providers/openlineage/transport/variable.py]: Leftover log.error?

**Review by kacpermuda — APPROVED** [review]
Looks good, leaving some nits - mostly naming or something similar.

File: providers/tests/system/openlineage/conftest.py (hunk 5 above)
Comment (mobuchowski) [reply providers/tests/system/openlineage/conftest.py:36]: Any test should take care of setting up listeners IMO. 
I don't think test instances should rely on some global side effects, like loading plugins because `import airflow` can result in that.

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 2 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: One place is generating key from Python classes, another one from dictionary... I think it's fine to have duplication factor == 2 in that case.

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 1 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: Added those.

Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: Fixed, thanks.

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 4 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: Fixed, thanks.

File: providers/src/airflow/providers/openlineage/transport/variable.py (hunk 6 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/transport/variable.py]: Removed.

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 3 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: Done.

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+            return False
+    elif expected != result:
+        log.error("Object of type %s: %s does not match %s", type(expected), expected, result)
+        return False
+    return True
+
+
+class OpenLineageTestOperator(BaseOperator):
+    """
+    This operator is added for system testing purposes.
```
Comment (dstandish) [inline providers/src/airflow/providers/openlineage/utils/operator.py]: if it is for testing purposes, why not put it in the tests folder tree?

Comment (dstandish) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: seems that helpers and tests and transports and such that exist only for tests, should be under tests/

File: airflow/models/variable.py [hunk 8]
```diff
@@ -152,6 +152,11 @@ def get(
                 mask_secret(var_val, key)
                 return var_val
 
+    @staticmethod
+    @provide_session
+    def list(session: Session = None):
```
Comment (Lee-W) [inline airflow/models/variable.py]:
```suggestion
    def list(session: Session = None) -> list[Variable]:
```

File: providers/src/airflow/providers/openlineage/transport/variable.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    """
+    This transport sends OpenLineage events to Variables.
+
+    Key schema is <DAG_ID>.<TASK_ID>.event.<EVENT_TYPE>.
+    It's made to be used in system tests, stored data read by OpenLineageTestOperator.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        ...
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/transport/variable.py]: Is this `...` needed?

File: providers/src/airflow/providers/openlineage/transport/variable.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+class VariableTransport(Transport, LoggingMixin):
+    """
+    This transport sends OpenLineage events to Variables.
+
+    Key schema is <DAG_ID>.<TASK_ID>.event.<EVENT_TYPE>.
+    It's made to be used in system tests, stored data read by OpenLineageTestOperator.
+    """
+
+    def __init__(self, *args, **kwargs):
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/transport/variable.py]:
```suggestion
    def __init__(self, *args, **kwargs) -> None:
```

File: providers/src/airflow/providers/openlineage/transport/variable.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        super().__init__(*args, **kwargs)
+        ...
+
+    def emit(self, event: Event):
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/transport/variable.py]:
```suggestion
    def emit(self, event: Event) -> None:
```

File: providers/src/airflow/providers/openlineage/transport/variable.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        key = f"{event.job.name}.event.{event.eventType.value.lower()}"  # type: ignore[union-attr]
+        event_str = Serde.to_json(event)
+        if (var := Variable.get(key, default_var=None, deserialize_json=True)) is not None:
+            Variable.set(key, value=var + [event_str], serialize_json=True)
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/transport/variable.py]:
```suggestion
            Variable.set(key=key, value=var + [event_str], serialize_json=True)
```

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+from airflow.models.variable import Variable
+
+if typing.TYPE_CHECKING:
+    from airflow.utils.context import Context
+
+log = logging.getLogger(__name__)
+
+
+def any(result):
+    return result
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
def any(result: Any) -> Any:
    return result
```

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+
+
+def is_datetime(result):
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
def is_datetime(result: Any) -> str:
```

Comment (Lee-W) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: I guess it's intentionally to be a string instead of a bool?

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    return "false"
+
+
+def is_uuid(result):
+    try:
+        uuid.UUID(result)
+        return "true"
+    except Exception:
+        pass
+    return "false"
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
def is_uuid(result: Any) -> bool:
    try:
        uuid.UUID(result)
        return "true"
    except Exception:
        pass
    return "false"
```

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        raise ValueError(f"Found {pattern} in {result}")
+    return "true"
+
+
+def url_scheme_authority(url) -> str:
+    parsed = urlparse(url)
+    return f"{parsed.scheme}://{parsed.netloc}"
+
+
+def url_path(url) -> str:
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
def url_path(url: str) -> str:
```

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+        raise ValueError(msg)
+
+
+def not_match(result, pattern) -> str:
+    if pattern in result:
+        raise ValueError(f"Found {pattern} in {result}")
+    return "true"
+
+
+def url_scheme_authority(url) -> str:
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
def url_scheme_authority(url: str) -> str:
```

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+    If there is no such environment variable set, return the default.
+    If the default is None, raise an exception for an undefined variable.
+    """
+    if var in os.environ:
+        return os.environ[var]
+    elif default is not None:
+        return default
+    else:
+        msg = f"Env var required but not provided: '{var}'"
+        raise ValueError(msg)
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
    if var in os.environ:
        return os.environ[var]
    elif default is not None:
        return default

    raise ValueError(f"Env var required but not provided: '{var}'")
```

Comment (Lee-W) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: I think we probably don't need that else block. the msg is not reused, probably also good to just pass it to the exception?

File: providers/src/airflow/providers/openlineage/utils/operator.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+            raise ValueError("Can't pass both event_templates and file_path")
+        if self.file_path is not None:
+            self.event_templates = {}
+            with open(file_path) as f:  # type: ignore[arg-type]
+                events = json.load(f)
+            for event in events:
+                key = event["job"]["name"] + ".event." + event["eventType"].lower()
+                self.event_templates[key] = event
+
+    def execute(self, context: Context):
```
Comment (Lee-W) [inline providers/src/airflow/providers/openlineage/utils/operator.py]:
```suggestion
    def execute(self, context: Context) -> None:
```

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 7 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: Moved those.

File: airflow/models/variable.py (hunk 8 above)
Comment (mobuchowski) [reply airflow/models/variable.py]: Fixed.

File: providers/src/airflow/providers/openlineage/utils/operator.py (hunk 14 above)
Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/utils/operator.py]: Yes - to be evaluated by Jinja comparison later.

**Comment by mobuchowski** [comment]
@Lee-W fixed all the typing/style comments.

File: airflow/models/variable.py [hunk 20]
```diff
@@ -152,6 +152,11 @@ def get(
                 mask_secret(var_val, key)
                 return var_val
 
+    @staticmethod
+    @provide_session
+    def list(session: Session = None) -> list[Variable]:
```
Comment (dstandish) [inline airflow/models/variable.py]: i'm not sure we want or need a method to list _all_ variables without any filtering. why is this needed?  i'm guessing you use it in a test or something.... if that's true, you should just write a query in the test.

Comment (mobuchowski) [reply airflow/models/variable.py]: Ok, removed.

File: providers/src/airflow/providers/openlineage/transport/variable.py [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+from openlineage.client.transport import Transport
+
+from airflow.models.variable import Variable
+from airflow.utils.log.logging_mixin import LoggingMixin
+
+if TYPE_CHECKING:
+    from openlineage.client.client import Event
+
+
+class VariableTransport(Transport, LoggingMixin):
```
Comment (dstandish) [inline providers/src/airflow/providers/openlineage/transport/variable.py]: if this is intended for system tests only, probably should be defined only in the tests subdir

**Review by dstandish — APPROVED** [review]
thanks @mobuchowski

File: providers/tests/system/openlineage/transport/variable.py [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+@register_transport
+class VariableTransport(Transport, LoggingMixin):
+    kind = "variable"
+    """
+    This transport sends OpenLineage events to Variables.
+
+    Key schema is <DAG_ID>.<TASK_ID>.event.<EVENT_TYPE>.
+    It's made to be used in system tests, stored data read by OpenLineageTestOperator.
+    """
```
Comment (Lee-W) [inline providers/tests/system/openlineage/transport/variable.py]:
```suggestion
    """
    This transport sends OpenLineage events to Variables.

    Key schema is <DAG_ID>.<TASK_ID>.event.<EVENT_TYPE>.
    It's made to be used in system tests, stored data read by OpenLineageTestOperator.
    """
    
    kind = "variable"
```

It looks like the docstring is for the class instead? if that's the case, probably reorder it this way would be better

**Review by Lee-W — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

File: providers/tests/system/openlineage/transport/variable.py (hunk 22 above)
Comment (mobuchowski) [reply providers/tests/system/openlineage/transport/variable.py]: Fixed.

File: providers/src/airflow/providers/openlineage/transport/variable.py (hunk 21 above)
Comment (potiuk) [reply providers/src/airflow/providers/openlineage/transport/variable.py]: Good point

Comment (mobuchowski) [reply providers/src/airflow/providers/openlineage/transport/variable.py]: already done :)

**Review by eladkal — APPROVED** [review]
